File: Algorithms-Comparisons-Java-Python/QuickSort.java
// QuickSort.java
// Implements both recursive and iterative versions of quicksort for integer arrays.

public class QuickSort {
    /**
     * Recursive quicksort implementation using Lomuto partition scheme.
     */
    public static void recursiveSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivot = partition(arr, low, high);
            recursiveSort(arr, low, pivot - 1);
            recursiveSort(arr, pivot + 1, high);
        }
    }

    /**
     * Iterative quicksort using an explicit stack to eliminate recursion.
     */
    public static void iterativeSort(int[] arr) {
        int n = arr.length;
        // stack for low/high indices
        int[] stack = new int[n];
        int top = -1;

        // push initial boundaries
        stack[++top] = 0;
        stack[++top] = n - 1;

        while (top >= 0) {
            int high = stack[top--];
            int low = stack[top--];

            if (low < high) {
                int pivot = partition(arr, low, high);
                // push right side
                if (pivot + 1 < high) {
                    stack[++top] = pivot + 1;
                    stack[++top] = high;
                }
                // push left side
                if (low < pivot - 1) {
                    stack[++top] = low;
                    stack[++top] = pivot - 1;
                }
            }
        }
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    private static void swap(int[] arr, int a, int b) {
        int tmp = arr[a];
        arr[a] = arr[b];
        arr[b] = tmp;
    }

    // simple utility main for manual testing
    public static void main(String[] args) {
        int[] data = {3,5,1,6,4,2};
        System.out.println("Before: ");
        for (int x : data) System.out.print(x + " ");
        System.out.println();
        recursiveSort(data, 0, data.length - 1);
        System.out.println("After recursive: ");
        for (int x : data) System.out.print(x + " ");
        System.out.println();
    }
}

File: QuickSort-Iterative-Recursive/QuickSort.cs
using System;

class Algorithm
{
    public static void Main(string[] args)
    {
        // Create sample data and display before sorting
        int[] numbers = { 33, 10, 55, 71, 29, 3, 18 };
        Console.WriteLine("Before sorting: " + string.Join(", ", numbers));

        // Perform quick sort on the entire array
        QuickSort(numbers, 0, numbers.Length - 1);

        // Show the results after sorting
        Console.WriteLine("After sorting:  " + string.Join(", ", numbers));
    }

    /// <summary>
    /// Sorts a segment of an integer array in place using the QuickSort algorithm.
    /// </summary>
    /// <param name="array">The array to be sorted.</param>
    /// <param name="low">Starting index of the segment to sort.</param>
    /// <param name="high">Ending index of the segment to sort.</param>
    public static void QuickSort(int[] array, int low, int high)
    {
        // Iterative version with tail recursion elimination to minimize stack usage.
        // Always sort the smaller partition first and loop on the larger one.
        while (low < high)
        {
            // Partition the segment and get pivot position
            int pivotIndex = Partition(array, low, high);

            // Determine which side is smaller
            int leftSize = pivotIndex - 1 - low;
            int rightSize = high - (pivotIndex + 1);

            if (leftSize < rightSize)
            {
                // Recurse on smaller left side
                QuickSort(array, low, pivotIndex - 1);
                // Update bounds to loop on larger right side
                low = pivotIndex + 1;
            }
            else
            {
                // Recurse on smaller right side
                QuickSort(array, pivotIndex + 1, high);
                // Update bounds to loop on larger left side
                high = pivotIndex - 1;
            }
        }
    }

    /// <summary>
    /// Partitions the array segment around a pivot using the Lomuto scheme.
    /// Values <= pivot move to the left; greater values stay on the right.
    /// Choosing the last element as pivot is simple but can degrade on sorted data.
    /// </summary>
    /// <returns>The final index of the pivot.</returns>
    private static int Partition(int[] array, int low, int high)
    {
        int pivot = array[high];      // choose last element as pivot
        int i = low - 1;              // place for swapping

        // Iterate over segment and move smaller values left of pivot
        for (int j = low; j < high; j++)
        {
            if (array[j] <= pivot)
            {
                i++;
                Swap(array, i, j);   // swap current element with i-th position
            }
        }

        // Place pivot after the last smaller element
        Swap(array, i + 1, high);
        return i + 1;
    }

    /// <summary>
    /// Swaps two elements in the given array.
    /// </summary>
    private static void Swap(int[] array, int a, int b)
    {
        int temp = array[a];
        array[a] = array[b];
        array[b] = temp;
    }
}